### ISSUE TITLE
search history?

### ISSUE THREAD

**Issue by CrazieIvan** [aaFn/Bookmark-search-plus-2#217]
Hi, thank you for this precious web extension, it's a life saver!

Would it be possible to add search history?  Many times I search the same words, and it would be really nice to click in the search field and get a list of the last several search words.  

Thanks very much for your work!

**Comment by aaFn** [comment]
Hello @CrazieIvan , I believe I can make a short list of **last 3** searches, which will be specific to each session / instance of BSP2 running = .e. if you have several windows / tabs with BSP2, each of them will have their own local history.

The history would last for the duration of the session = i.e., when you close the instance, local history is forgotten, and when you restart FF, or start a new BSP2 session, the history is empty ans starts populating.

**Comment by aaFn** [comment]
This is in 2.0.102 which is out now.
Let me know if any problem, aaFn.

**Comment by CrazieIvan** [comment]
Oh crap, I just saw this.  :-)
OK I tried it, and I think it works as intended by design, but I'm not sure it's 100% usable in its current state.

1)  The problem is that when I type a word slowly, BSP2 auto-searches the partial words before i'm finished typing the full word.  So the history remembers all the intermediate searches for the partial words.
So my history list would show parts of the same word search, like:
"goog"
"googl"
"google"

I hope I'm explaining this well enough. Does it make sense?  

2)  Also, assuming you can figure out how to fix the above issue, would it be possible to increase the list because 3 items is very limited?  Maybe 10 or 15 would be nice.  

3)  Would it be possible to keep a shared history list for the current FF session between all the windows and tabs?

I appreciate your efforts!

**Comment by aaFn** [comment]
Hello @CrazieIvan :

- On 1), we can decide to keep the most specific or the least specific, instead of multiplying, as you want ..
- On 2), too big a number is also no good, because of size of the "selection drop down" .. can easily go to 5 for now, and then we can see function of 1)
- On 3), this is maintained in local HTML objects, and not easy to share. The complexity of it is not realy worth it. In addition, this allows globally to remember more values across all instances, which is an advantage and also an answer to 2).
In short, doing 3) would hit more severely on 2) and would not bring big advantages to compensate, but would be significant work.

**Comment by CrazieIvan** [comment]
@aaFn 
Thank you for the quick response :-)

1) Sorry, I don't understand what you mean by "most specific" or "least specific".  
2)  OK, I understand.  I'm happy with whatever is possible then.
3) If it's too challenging, then don't worry about it.  It would be nice if history is shared between windows, but if not, then it's still plenty good if it's shared within the same window (which it seems to be right now).

**Comment by aaFn** [comment]
On 1), "google" is more specific than "goog" .. so would you want to keep "google" (most specific) or "goog" (least specific) in the list of remembered searches ?

On 2), ok, I'll change to 5.

Thanks, aaFn.

**Comment by CrazieIvan** [comment]
I see.  Well, ideally it would remember the full word that I type, so in that example it would be 'google'.  I just wonder, how would it know when I'm done typing in order to know what expression to commit to memory?  Is it based on time duration between key presses?

**Comment by aaFn** [comment]
Simply, each time a search is triggered (after timeour on key typing), then the value is stored in history.
While storing, I can check if that includes (= is more specific than) a previous saved string, and then replace the previous saved one by the more specific one.

**Comment by CrazieIvan** [comment]
Oh so you are comparing the latest search to a list of intermediate prior searches.  That makes sense.
